### PR TITLE
fix: don't duplicate existing filters

### LINKED ISSUE #408: Optimising an already optimised logical plan results in redundant pushed down filters
**Describe the bug**

If you run a `LogicalPlan` that has a `TableProviderFilterPushDown` `TableProvider`, then the filter pushdown optimiser rule will keep adding filters to the pushed down table scan without checking if they're already present.

Given the plan:

```
Filter: #a Eq Int64(1)
  TableScan: projection=None
```

it will be optimised to the following plan the first time it runs through the filter:

```
Filter: #a Eq Int64(1)
      TableScan: projection=None, filters=[#a Eq Int64(1)]
```

If you run it through the optimiser again then it will not be left as is. Instead it will become:

```
Filter: #a Eq Int64(1)
  TableScan: projection=None, filters=[#a Eq Int64(1), #a Eq Int64(1)]
```

**To Reproduce**

Run a logical plan that pushes predicates down to the table provider through the plan optimiser twice. All pushed down expression filters will be duplicated in the table  scan filters.

**Expected behavior**

The plan should not change if it's ran through the optimiser multiple times.

**Additional context**

Whilst the current behaviour produces correct plans they are not optimal because they can contain redunant duplicate filters if they're run through the optimiser multiple times.

### PR DESCRIPTION
**Pull request by e-dard**
# Which issue does this PR close?

Closes #408

# What changes are included in this PR?

This PR ensures that filter expressions already present on the table scan part of a logical plan are not re-added if the plan is run through the optimiser again.

### CODE REVIEW

File: datafusion/src/optimizer/filter_push_down.rs [hunk 1]
```diff
@@ -1037,6 +1045,23 @@ mod tests {
...
+
+        let optimised_plan = optimize_plan(&plan);
+
+        let expected = "\
+        Filter: #a Eq Int64(1)\
+        \n  TableScan: projection=None, filters=[#a Eq Int64(1)]";
+
+        // Optimizing the same plan multiple times should produce the same plan
+        // each time.
+        assert_optimized_plan_eq(&optimised_plan, expected);
```
Comment (alamb) [inline datafusion/src/optimizer/filter_push_down.rs:1061]: 👍

File: datafusion/src/optimizer/filter_push_down.rs [hunk 2]
```diff
@@ -379,6 +379,11 @@ fn optimize(plan: &LogicalPlan, mut state: State) -> Result<LogicalPlan> {
                 }
 
                 if add_to_provider {
+                    // Don't add expression again if it's already present in
+                    // pushed down filters.
+                    if new_filters.iter().any(|filter| filter.eq(filter_expr)) {
```
Comment (Dandandan) [inline datafusion/src/optimizer/filter_push_down.rs:384]:
```suggestion
                    if new_filters.contains(filter_expr) {
```
something like this should work?

**Review by alamb — APPROVED** [review]
Looks good to me . Thank you @e-dard 

FYI @jorgecarleitao

File: datafusion/src/optimizer/filter_push_down.rs (hunk 1 above)
Comment (NGA-TRAN) [reply datafusion/src/optimizer/filter_push_down.rs:1061]: +1

File: datafusion/src/optimizer/filter_push_down.rs (hunk 2 above)
Comment (NGA-TRAN) [reply datafusion/src/optimizer/filter_push_down.rs:384]: Nice

**Review by NGA-TRAN — APPROVED** [review]

File: datafusion/src/optimizer/filter_push_down.rs (hunk 2 above)
Comment (e-dard) [reply datafusion/src/optimizer/filter_push_down.rs]: Thanks @Dandandan, I completely forgot about `contains` on `Vec`.

**Review by Dandandan — APPROVED** [review]

**Pull request merged**